internal/utils/nat.go

```diff
@@ -0,0 +1,168 @@
+	if err != nil {
+		return nil, err
+	}
+	time.Sleep(5 * time.Second)
+	for lanId, gatewayIps := range lansToGateways {
+		id, err := strconv.ParseInt(lanId, 10, 32)
+		if err != nil {
+			return nil, err
+		}
+		// Unpack the map into NatGatewayLanProperties objects. https://api.ionos.com/docs/cloud/v6/#tag/NAT-Gateways/operation/datacentersNatgatewaysPost
+		var ptrGatewayIps *[]string = nil
+		if len(gatewayIps) > 1 || gatewayIps[0] != "" {
+			// We do this check so that we don't set the GatewayIps property if it's empty. If the property is empty, a gateway IP is generated by the API.
+			ptrGatewayIps = &gatewayIps
+		}
+		lans = append(lans, sdkgo.NatGatewayLanProperties{Id: pointer.To(int32(id)), GatewayIps: ptrGatewayIps})
+	}
+
+	nrm := NewNRM(publicIps[0], subnet, subnet)
+	nrm.
+		OpenPort("TCP", 22).            // SSH
+		OpenPort("UDP", 53).            // DNS
+		OpenPort("TCP", 80).            // HTTP
+		OpenPort("TCP", 179).           // Calico BGP Port
+		OpenPort("TCP", 443).           //
+		OpenPort("TCP", 2376).          // Node driver Docker daemon TLS port
+		OpenPort("UDP", 4789).          // Flannel VXLAN overlay networking on Windows cluster
+		OpenPort("TCP", 6443).          // Rancher Webhook
+		OpenPort("TCP", 6783).          // Weave Port
+		OpenPort("TCP", 8443).          // Rancher webhook
+		OpenPort("UDP", 8472).          // Canal/Flannel VXLAN overlay networking
+		OpenPort("TCP", 9099).          // Canal/Flannel livenessProbe/readinessProbe
+		OpenPort("TCP", 9100).          // Default port required by Monitoring to scrape metrics from Linux node-exporters
+		OpenPort("TCP", 9443).          // Rancher webhook
+		OpenPort("TCP", 9796).          // Default port required by Monitoring to scrape metrics from Windows node-exporters
+		OpenPort("TCP", 10254).         // Ingress controller livenessProbe/readinessProbe
+		OpenPort("TCP", 10256).         //
+		OpenPorts("TCP", 2379, 2380).   // etcd
+		OpenPorts("UDP", 6783, 6784).   // Weave Port (UDP)
+		OpenPorts("TCP", 10250, 10252). // Metrics server communication with all nodes API
+		OpenPorts("TCP", 30000, 32767). //
+		OpenPorts("UDP", 30000, 32767). //
+		OpenPort("ALL", 0)              // Outbound
+	rules := nrm.Make()
+
+	nat, resp, err := c.NATGatewaysApi.DatacentersNatgatewaysPost(c.ctx, datacenterId).NatGateway(
+		sdkgo.NatGateway{
+			Properties: &sdkgo.NatGatewayProperties{
+				Name:      &name,
+				PublicIps: &publicIps,
+				Lans:      &lans,
+			},
+			Entities: &sdkgo.NatGatewayEntities{
+				Rules:    &sdkgo.NatGatewayRules{Items: rules},
+				Flowlogs: nil,
+			},
+		},
+	).Execute()
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.waitTillProvisioned(resp.Header.Get("location"))
+	return &nat, err
+}
+
+func (c *Client) createLansIfNotExist(datacenterId string, lanIds []string) error {
+	for _, lanid := range lanIds {
+		_, resp, err := c.LANsApi.DatacentersLansFindById(c.ctx, datacenterId, lanid).Execute()
+		if resp.StatusCode == 404 {
```

Comment:
This can crash because Statuscode is in *http.Response and either resp, or resp.Response can be nil. We had crashes like this in terraform. Use resp.HttpNotFound() function instead.

Reply:
thanks so much!